### PR TITLE
BuildResidentialHPXML measure: Rim joists

### PR DESCRIPTION
**Pull request by joseph-robertson**
## Pull Request Description

Add arguments:
* `geometry_rim_joist_height`
* `rim_joist_assembly_r`

Remove arguments:
* `wall_solar_absorptance`
* `wall_emittance`
* `roof_solar_absorptance`
* `roof_emittance`

Default arguments:
* `roof_color` to `HPXML::ColorMedium`
* `wall_color` to `HPXML::ColorMedium`

Update argument descriptions:
* `wall_siding_type` applies to rim joists
* `wall_color` applies to rim joists

Update `Geometry` methods to create rim joist surfaces:
* `create_single_family_detached`
* `create_single_family_attached`
* `create_multifamily`

Re-enable `rim_joists` check in test file. Disable `solar_absorptance`, `emittance`, and `color` checks in test file.

Changes to test xml files:
* `base-enclosure-2stories-garage.xml`
  * Increase foundation wall area from 880 to 1200 (same as `base-enclosure-garage.xml`)
* `base-bldgtype-single-family-attached.xml`
  * Decrease rim joist adjacent to outside's area from 116 to 66 (116 was SFD's rim joist area)
  * Add another rim joist adjacent to basement with area 28
* Change insulation assembly R-value from 2.3 to 4.0 for uninsulated rim joists

## Checklist

Not all may apply:

- [ ] EPvalidator.xml has been updated
- [ ] Tests (and test files) have been updated
- [ ] Documentation has been updated
- [ ] Changelog has been updated
- [ ] `openstudio tasks.rb update_measures` has been run
- [ ] No unexpected regression test changes on CI

### CODE REVIEW

File: BuildResidentialHPXML/measure.rb [hunk 1]
```diff
@@ -345,6 +345,13 @@ def arguments(model)
     arg.setDefaultValue(0.0)
     args << arg
 
+    arg = OpenStudio::Measure::OSArgument::makeDoubleArgument('geometry_rim_joist_height', true)
+    arg.setDisplayName('Geometry: Rim Joist Height')
+    arg.setUnits('ft')
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:350]: Let's make this units of inches so someone doesn't have to convert 7.25 inches (2x8) to feet.

File: BuildResidentialHPXML/measure.rb [hunk 2]
```diff
@@ -345,6 +345,13 @@ def arguments(model)
...
+    arg.setDescription("The height of the rim joists. Only applies to basements/crawlspaces. Only applies to #{HPXML::ResidentialTypeSFD} and #{HPXML::ResidentialTypeSFA} units.")
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:351]: Why does this not apply to MF units? Your other new argument doesn't have a similar statement, and I would think that it should apply to a MF unit over a crawlspace.

File: BuildResidentialHPXML/measure.rb [hunk 3]
```diff
@@ -345,6 +345,13 @@ def arguments(model)
...
+    arg.setDefaultValue(0.0)
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:352]: Let's default to 5.5.

File: BuildResidentialHPXML/measure.rb [hunk 4]
```diff
@@ -493,6 +500,12 @@ def arguments(model)
     arg.setDefaultValue(Constants.Auto)
     args << arg
 
+    arg = OpenStudio::Measure::OSArgument::makeDoubleArgument('rim_joist_assembly_r', false)
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:503]: Why is this argument not required but the other one is?

File: BuildResidentialHPXML/measure.rb [hunk 5]
```diff
@@ -3037,6 +3050,10 @@ def validate_arguments(args)
       errors << "foundation_wall_insulation_distance_to_bottom=#{args[:foundation_wall_insulation_distance_to_bottom]} and geometry_foundation_height=#{args[:geometry_foundation_height]}" if error
     end
 
+    # ambient with rim joists
+    error = (args[:geometry_foundation_type] == HPXML::FoundationTypeAmbient) && (args[:geometry_rim_joist_height] > 0)
+    errors << "geometry_foundation_type=#{args[:geometry_foundation_type]} and geometry_rim_joist_height=#{args[:geometry_rim_joist_height]}" if error
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:3055]: It also won't apply to a slab foundation type. But I'm tempted to say not to generate an error here and just ignore it if it doesn't apply (especially if we have a default > 0).

File: BuildResidentialHPXML/measure.rb [hunk 6]
```diff
@@ -3467,10 +3483,64 @@ def self.set_roofs(hpxml, runner, model, args)
...
+        insulation_assembly_r_value = args[:rim_joist_assembly_r]
+      end
+
+      if exterior_adjacent_to == HPXML::LocationOutside && args[:wall_siding_type].is_initialized
+        siding = args[:wall_siding_type].get
+      end
+
+      if args[:wall_color] == Constants.Auto && args[:wall_solar_absorptance] == Constants.Auto
+        solar_absorptance = 0.7
+      end
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:3504]: This is weird, why are we hard-coding a value here?

File: tasks.rb [hunk 7]
```diff
@@ -514,6 +515,7 @@ def get_values(osw_file, step)
     step.setArgument('geometry_foundation_type', HPXML::FoundationTypeBasementConditioned)
     step.setArgument('geometry_foundation_height', 8.0)
     step.setArgument('geometry_foundation_height_above_grade', 1.0)
+    step.setArgument('geometry_rim_joist_height', 0.78)
```
Comment (shorowit) [inline tasks.rb:518]: Something is not right if the rim joist height needs to be 0.78*12=9.36 inches.

File: tasks.rb [hunk 8]
```diff
@@ -3501,6 +3511,8 @@ def set_hpxml_rim_joists(hpxml_file, hpxml)
                          solar_absorptance: 0.7,
                          emittance: 0.92,
                          insulation_assembly_r_value: 23.0)
+  elsif ['base-bldgtype-single-family-attached.xml'].include? hpxml_file
+    hpxml.rim_joists[-1].area = 67
```
Comment (shorowit) [inline tasks.rb:3515]: If we are changing the sample files, that should be done in a separate PR against master.

File: BuildResidentialHPXML/tests/build_residential_hpxml_test.rb [hunk 9]
```diff
@@ -207,16 +212,22 @@ def _check_hpxmls(workflow_dir, built_dir, test_dir, hpxml_path)
...
+      hpxml.attics.clear
       hpxml.foundation_walls.each do |foundation_wall|
         next if foundation_wall.insulation_assembly_r_value.nil?
         foundation_wall.insulation_assembly_r_value = foundation_wall.insulation_assembly_r_value.round(2)
       end
+      if hpxml.rim_joists.length > 0
+        (0...hpxml.rim_joists.length).to_a.reverse.each do |i|
+          next unless [HPXML::LocationLivingSpace].include? hpxml.rim_joists[i].interior_adjacent_to
+
+          hpxml.rim_joists.delete_at(i)
```
Comment (shorowit) [inline BuildResidentialHPXML/tests/build_residential_hpxml_test.rb:225]: Why are we having to do this?

**Review by shorowit — COMMENTED** [review]
Some initial comments.

File: BuildResidentialHPXML/measure.rb (hunk 4 above)
Comment (joseph-robertson) [reply BuildResidentialHPXML/measure.rb:503]: I was following `foundation_wall_assembly_r`, but I suppose I shouldn't have been. I'll make it required, and default to 23.

File: tasks.rb (hunk 7 above)
Comment (joseph-robertson) [reply tasks.rb:518]: This is why I was asking about the dimensions of `base.xml`. At 30 x 45 (perimeter of 150), the rim joist would need to be 0.78 inches (0.78 x 150)  so that the rim joist area is 116 ft^2.

File: BuildResidentialHPXML/tests/build_residential_hpxml_test.rb (hunk 9 above)
Comment (joseph-robertson) [reply BuildResidentialHPXML/tests/build_residential_hpxml_test.rb:225]: To exclude any rim joists between living space and outdoors (i.e., we decided not to handle 2nd story rim joists with the build measure).

File: tasks.rb (hunk 8 above)
Comment (joseph-robertson) [reply tasks.rb:3515]: Yeah, I'll make it there too if you deem these are good changes.

File: BuildResidentialHPXML/measure.rb (hunk 2 above)
Comment (joseph-robertson) [reply BuildResidentialHPXML/measure.rb:351]: You're right. I think I forgot to handle MF because there are no test xml files with units adjacent to foundations.

File: BuildResidentialHPXML/measure.rb (hunk 6 above)
Comment (joseph-robertson) [reply BuildResidentialHPXML/measure.rb:3504]: 🤷‍♂️

File: workflow/sample_files/base-enclosure-2stories-garage.xml [hunk 10]
```diff
@@ -207,7 +207,7 @@
             <ExteriorAdjacentTo>ground</ExteriorAdjacentTo>
             <InteriorAdjacentTo>basement - conditioned</InteriorAdjacentTo>
             <Height>8.0</Height>
-            <Area>880.0</Area>
+            <Area>1200.0</Area>
```
Comment (shorowit) [inline workflow/sample_files/base-enclosure-2stories-garage.xml:210]: Good change, this does look right. Seems like it was using the _exposed_ perimeter of the basement to calculate the area (110x8=880) instead of the _total_ perimeter (150x8=1200).

File: BuildResidentialHPXML/measure.rb [hunk 11]
```diff
@@ -3467,8 +3484,67 @@ def self.set_roofs(hpxml, runner, model, args)
...
+
+      hpxml.rim_joists.add(id: valid_attr(surface.name),
+                           exterior_adjacent_to: exterior_adjacent_to,
+                           interior_adjacent_to: interior_adjacent_to,
+                           area: UnitConversions.convert(surface.grossArea, 'm^2', 'ft^2').round(1),
+                           siding: siding,
+                           color: color,
+                           solar_absorptance: solar_absorptance,
+                           emittance: emittance,
+                           insulation_assembly_r_value: args[:rim_joist_assembly_r])
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:3537]: We should only assign this R-value to thermal boundary rim joists. So do the same thing as we do for walls [here](https://github.com/NREL/OpenStudio-HPXML/blob/87e7a869e84be16b000e16b1cf602f6a2ffe17fc/BuildResidentialHPXML/measure.rb#L3645-L3649) (excluding the `is_uncond_attic_roof_insulated` part, since we're not dealing with attic gable walls).

File: workflow/sample_files/base-bldgtype-single-family-attached.xml [hunk 12]
```diff
@@ -122,6 +122,18 @@
...
+          <RimJoist>
+            <SystemIdentifier id='RimJoistOther'/>
+            <ExteriorAdjacentTo>basement - conditioned</ExteriorAdjacentTo>
+            <InteriorAdjacentTo>basement - conditioned</InteriorAdjacentTo>
+            <Area>28.0</Area>
+            <SolarAbsorptance>0.7</SolarAbsorptance>
+            <Emittance>0.92</Emittance>
+            <Insulation>
+              <SystemIdentifier id='RimJoistOtherInsulation'/>
+              <AssemblyEffectiveRValue>23.0</AssemblyEffectiveRValue>
```
Comment (shorowit) [inline workflow/sample_files/base-bldgtype-single-family-attached.xml:134]: If you make the previous change correctly, this should change to R-4.0 (while the other rim joist remains R-23.0).

File: BuildResidentialHPXML/measure.rb [hunk 13]
```diff
@@ -345,6 +345,13 @@ def arguments(model)
     arg.setDefaultValue(0.0)
     args << arg
 
+    arg = OpenStudio::Measure::OSArgument::makeDoubleArgument('geometry_rim_joist_height', true)
+    arg.setDisplayName('Geometry: Rim Joist Height')
+    arg.setUnits('in')
+    arg.setDescription('The height of the rim joists. Only applies to basements/crawlspaces.')
+    arg.setDefaultValue(5.5)
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:352]: In hindsight, let's make this default to 9.25.

File: tasks.rb (hunk 7 above)
Comment (shorowit) [reply tasks.rb]: Okay, this is resolved. The rim joist was 9.25 inches (2x10), but the area was rounded making me think it was 9.36 inches.

**Review by shorowit — APPROVED** [review]
Looks good. And a net reduction in arguments!
